src/typechecker.rs

```diff
@@ -2611,38 +2603,29 @@ fn typecheck_function_predecl(
             variable: checked_variable.clone(),
         });
 
-        if let Some(check_scope) = check_scope {
-            let _ = project.add_var_to_scope(check_scope, checked_variable, param.variable.span);
-        }
+        let _ = project.add_var_to_scope(check_scope, checked_variable, param.variable.span);
     }
 
     let (function_return_type_id, err) =
         typecheck_typename(&function.return_type, function_scope_id, project);
     error = error.or(err);
 
-    checked_function.return_type_id = function_return_type_id;
+    let (block, _) = typecheck_block(&function.block, check_scope, project, SafetyMode::Safe);
```

Comment:
We don't want to typecheck the block in the `predecl` version of the function, as we won't have all the types in scope yet. The `predecl` part currently is intended only to get the names and types ready for the full check which is done in a second part of the typechecking phase.

Reply:
Yeah, but typechecking the block is the only way to infer the type of the fat arrow function

Reply:
I mean you could, but that feels like half a fix that will fail in surprising ways in practice. Unless you check in the order that respects the dependencies of functions and types on each other, you can't start checking the bodies before we know what is in scope.

Here's an idea: do all the predecls first. For any function that has an `Unknown` as its output, check those bodies first. Then do the full typecheck of everything else. Basically one pass for the predecl (like we do now), one pass specifically for inferring return types, and then one pass to finish the typecheck (like we do now)

Reply:
oh okay

Reply:
maybe I should just move the typechecking of the block inside the if expression where the return type is inferred from the block because it had no hints.

Reply:
Okay, I made the block typecheck only run when the function return type is `UNKNOWN`, although what you're saying of a second pass makes more sense. On it!